include/AudioPluginInterface.h

```diff
@@ -0,0 +1,418 @@
+ *
+ * Copyright (c) 2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_AUDIO_PLUGIN_INTERFACE_H
+#define LMMS_AUDIO_PLUGIN_INTERFACE_H
+
+#include "AudioData.h"
+#include "AudioPluginBuffer.h"
+#include "Effect.h"
+#include "Instrument.h"
+#include "InstrumentTrack.h"
+#include "PluginPinConnector.h"
+#include "SampleFrame.h"
+
+namespace lmms
+{
+
+enum class ProcessStatus
+{
+	//! Unconditionally continue processing
+	Continue,
+
+	//! Calculate the RMS out sum and call `checkGate` to determine whether to stop processing
+	ContinueIfNotQuiet,
+
+	//! Do not continue processing
+	Sleep
+};
+
+//! Compile time customizations for `AudioPluginInterface` to meet the needs of a plugin implementation
+struct PluginConfig
+{
+	//! The audio data layout used by the plugin
+	AudioDataLayout layout;
+
+	//! The number of plugin input channels, or `DynamicChannelCount` if unknown at compile time
+	int inputs = DynamicChannelCount;
+
+	//! The number of plugin output channels, or `DynamicChannelCount` if unknown at compile time
+	int outputs = DynamicChannelCount;
+
+	//! In-place processing - true (always in-place) or false (dynamic, customizable in audio buffer impl)
+	bool inplace = false;
+
+	//! If true, plugin implementation will provide an `AudioPluginBufferInterfaceProvider`
+	bool customBuffer = false;
+};
+
+class NotePlayHandle;
+
+namespace detail
+{
+
+//! Provides the correct `processImpl` interface for instruments or effects to implement
+template<class ChildT, typename BufferT, typename ConstBufferT, bool inplace, bool customBuffer>
+class AudioProcessingMethod;
+
+//! Instrument specialization
+template<typename BufferT, typename ConstBufferT>
+class AudioProcessingMethod<Instrument, BufferT, ConstBufferT, false, false>
+{
+protected:
+	//! The main audio processing method for NotePlayHandle-based Instruments
+	//! NOTE: NotePlayHandle-based instruments are currently unsupported
+	//virtual void processImpl(NotePlayHandle* nph, ConstBufferT in, BufferT out) {}
+
+	//! The main audio processing method for MIDI-based Instruments
+	virtual void processImpl(ConstBufferT in, BufferT out) = 0;
```

Comment:
The `in` is for instruments with sidechain inputs. If the instrument doesn't have sidechain inputs, `in` will be an empty buffer.

Reply:
From all I recall, LMMS does not at all support plugins with sidechain inputs. Are you talking about (yet) unreachable code?

Reply:
Sidechain inputs fall into the category of multiple plugin channels, and multiple plugin channels are supported by the pin connector and currently implemented by both VST instruments and VST effects. However since without audio routing there is no way to route a sidechain signal to an instrument's sidechain input, an instrument's sidechain input will always be zeroed. So the functionality is there but not really usable yet.